Peer discovery with Kademlia protocol (#100)
libp2p/cpp-libp2p: example/02-kademlia/README.md

# Example using Kademlia for peer discovery

## General description

* `rendezvous_chat.cpp` is a simple chat-client with peer discovery in DHT

## Launching

### rendezvous_chat

`./rendezvous_chat [LISTEN_ADDR]`

where `LISTEN_ADDR` is libp2p address for incoming connection

Explanation by example: `./rendezvous_chat /ip4/127.0.0.1/tcp/3333` means that

> * local node will listen to port 3333

Reviewer: Is this "*" needed here?

Author: I guess is not. Thanks.

# Example using Kademlia for peer discovery

## General description

* `rendezvous_chat.cpp` is a simple chat-client with peer discovery in DHT

## Launching

### rendezvous_chat

```bash
./rendezvous_chat [LISTEN_ADDR]
```

where `LISTEN_ADDR` is libp2p address for incoming connection

Explanation by example: 
`./rendezvous_chat /ip4/127.0.0.1/tcp/3333` means that local node will listen port 3333 on IP 127.0.0.1

### What is happening after launching

> 1. app connects to botstrap nodes and start random-walking mechanism to collect peers info

Reviewer 2: ```suggestion
1. app connects to bootstrap nodes and starts random-walking mechanism to collect peer infos
```

> 1. noticing network (known peers) about owns providing of chat features

Reviewer 2: ```suggestion
2. The network (known peers) is notified about owns providing of chat features
```
To be honest, the second part is not very clear.

> 2. The network (known peers) is notified about owns providing of chat features

Reviewer 3: what is chat feature? maybe chat protocol?

> 2. The network (known peers) is notified about owns providing of chat features

Author: Agree, "chat protocol" is closest term.

> 1. inding others providers with the same chat features

Reviewer 2: ```suggestion
3. The app starts searching for others providers with the same chat features
```

> 1. connecting to found peers and acception connections of others

Reviewer 2: ```suggestion
1. The app connects to the found peers and accepts incoming connections
```

> 4. The app connects to the found peers and accepts incoming connectionsThe app connects to the found peers and accepts incoming connections

Reviewer 3: ```suggestion
4. The app connects to the found peers and accepts incoming connections. The app connects to the found peers and accepts incoming connections
```

### Normal work
First message in terminal contains own peer_id:
```log
12D3KooWKpQQojSESeYK5wgxhiHFfWGt1DT37WMks4e2wXjjTrZj * started
```

> Incoming and outgoing connections looks like:

Reviewer 2: ```suggestion
Incoming and outgoing connections look like:
```

```log
12D3KooWNrhWQqbFkwg3Viz3HL2jCdvK9Q3Nieu9QXX37c17Aw2Y + outgoing stream to /ip4/127.0.0.1/tcp/3333
12D3KooWN4U92XG49kSjJ6p5TUEWxj76V432HfbmQAXwryLApkue + incoming stream from /ip4/127.0.0.1/tcp/58442
```

> After connection was established you cant start chatting by type message in terminal:

Reviewer 2: ```suggestion
After connection was established you can start chatting by typing a message in terminal:
```

```log

> Hello! Are you from?
> 12D3KooWN4U92XG49kSjJ6p5TUEWxj76V432HfbmQAXwryLApkue < Hello! Are you from?

Reviewer 2: ```suggestion
Hello! Where are you from?
12D3KooWN4U92XG49kSjJ6p5TUEWxj76V432HfbmQAXwryLApkue < Hello! Where are you from?
```
